consensus/state.go

```diff
@@ -1836,45 +1850,65 @@ func (cs *State) addProposalBlockPart(msg *BlockPartMessage, peerID p2p.ID) (add
 	}
 
 	// We're not expecting a block part.
-	if cs.ProposalBlockParts == nil {
+	if cs.ProposalBlockRows == nil {
 		// NOTE: this can happen when we've gone to a higher round and
 		// then receive parts from the previous round - not necessarily a bad peer.
 		cs.Logger.Info("Received a block part when we're not expecting any",
-			"height", height, "round", round, "index", part.Index, "peer", peerID)
+			"height", height, "round", round, "index", "peer", peerID)
 		return false, nil
 	}
+	if cs.ProposalBlockRows.TotalSize() > int(cs.state.ConsensusParams.Block.MaxBytes) {
+		return false, fmt.Errorf("got row of a proposal block exceeding maximum block size (%d > %d)",
+			cs.ProposalBlockRows.TotalSize(), cs.state.ConsensusParams.Block.MaxBytes,
+		)
+	}
 
-	added, err = cs.ProposalBlockParts.AddPart(part)
+	added, err = cs.ProposalBlockRows.AddRow(part)
 	if err != nil {
 		return added, err
 	}
-	if cs.ProposalBlockParts.ByteSize() > cs.state.ConsensusParams.Block.MaxBytes {
-		return added, fmt.Errorf("total size of proposal block parts exceeds maximum block bytes (%d > %d)",
-			cs.ProposalBlockParts.ByteSize(), cs.state.ConsensusParams.Block.MaxBytes,
-		)
-	}
-	if added && cs.ProposalBlockParts.IsComplete() {
-		bz, err := ioutil.ReadAll(cs.ProposalBlockParts.GetReader())
+	if added && cs.ProposalBlockRows.IsComplete() {
+		square, err := cs.ProposalBlockRows.Square()
 		if err != nil {
 			return added, err
 		}
 
-		var pbb = new(tmproto.Block)
-		err = proto.Unmarshal(bz, pbb)
+		data, err := types.DataFromSquare(square)
 		if err != nil {
 			return added, err
 		}
 
-		block, err := types.BlockFromProto(pbb)
-		if err != nil {
-			return added, err
+		var commit *types.Commit
+		switch {
+		case cs.Height == cs.state.InitialHeight:
+			// We're creating a proposal for the first block.
+			// The commit is empty, but not nil.
+			commit = types.NewCommit(0, 0, types.BlockID{}, nil)
+		case cs.LastCommit.HasTwoThirdsMajority():
+			// Make the commit from LastCommit
+			commit = cs.LastCommit.MakeCommit()
+		default: // This shouldn't happen.
+			return added, fmt.Errorf("no commit for the previous block")
 		}
 
-		cs.ProposalBlock = block
-		cs.ProposalBlockRows, err = block.RowSet(context.TODO(), cs.dag)
+		cs.ProposalBlock = cs.state.MakeBlock(
+			cs.Proposal.Height,
+			data.Txs,
+			data.Evidence.Evidence,
+			data.IntermediateStateRoots.RawRootsList,
+			data.Messages,
+			commit,
+			cs.Validators.GetProposer().Address,
+		)
+
+		// TODO(Wondertan): This is unnecessary in general, but for now it writes needed fields
+		//  and specifically NumOriginalDataShares, which likely should be par of the proposal
+		cs.ProposalBlockRows, err = cs.ProposalBlock.RowSet(context.TODO(), mdutils.Mock())
 		if err != nil {
-			return false, err
+			return added, err
 		}
+		cs.ProposalBlockParts = cs.ProposalBlock.MakePartSet(types.BlockPartSizeBytes)
+
```

Comment:
This part is important to be reviewed. We were discussing that the Proposal needs to have LastCommit if row gossiping is implemented, but the node can also make it itself, is this correct?

Reply:
Also, I see a reason to add `NumOriginalDataShares` to `Proposal`, otherwise, validators would need to `ComputeShares` themselves to get the number and add it to `Header`.